### PR TITLE
Suport node and IP discovery without kvstore

### PR DESCRIPTION
**Pull request by tgraf**
Now that the identity allocation is supported via CRD, the remaining blockers to enable kvstore free operation is node and ipcache discovery.

This PR fixes all remaining issues to run Cilium without a kvstore by adding support for both node and ipcache discovery via existing Kubernetes custom resources.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8627)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/k8s-1.10-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:420
unable to apply https://raw.githubusercontent.com/cilium/star-wars-demo/v1.0/01-deathstar.yaml: 
Expected command: kubectl apply -f  https://raw.githubusercontent.com/cilium/star-wars-demo/v1.0/01-deathstar.yaml 
To succeed, but it failed:
Exitcode: 1 
Stdout:
 	 
Stderr:
```

Looks like it hit #8634

**Comment by tgraf** [comment]
test-me-please

**Comment by ianvernon** [comment]
@tgraf https://github.com/cilium/cilium/pull/8627#issuecomment-513896175 could have occurred due to connectivity issues trying to get the resource from GitHub, as the call to `kubectl apply` in the `demos.go` test has a timeout of 30 seconds to apply to resource, which is a reasonable amount of time. I suspect this may be due to GitHub having disruption of service earlier today.

File: pkg/k8s/client_test.go [hunk 1]
```diff
@@ -147,7 +148,7 @@ func (s *K8sSuite) TestUseNodeCIDR(c *C) {
 		})
 
 	node2Slim := ConvertToNode(node2.DeepCopy()).(*types.Node)
-	node2Cilium := ParseNode(node2Slim, node.FromAgentLocal)
+	node2Cilium := ParseNode(node2Slim, source.Unspec)
```
Comment (ianvernon) [inline pkg/k8s/client_test.go:151]: Why the change in source? Before, the changes were due to refactors, this looks like a functional change reflected in the unit tests.

File: pkg/k8s/init.go [hunk 2]
```diff
@@ -82,7 +83,9 @@ func retrieveNodeInformation(nodeName string) (*node.Node, error) {
 	}
 	typesNode := nodeInterface.(*types.Node)
 
-	n := ParseNode(typesNode, node.FromAgentLocal)
+	// The source is left unspecified as this node resource should never be
+	// used to update state
+	n := ParseNode(typesNode, source.Unspec)
```
Comment (ianvernon) [inline pkg/k8s/init.go:88]: Why does this never need to be used to update state as opposed to before? Was it never used to update state? This may answer my question above in the unit tests, I'm guessing.

File: pkg/source/source.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	// overwritten by all other sources
+	Generated Source = "generated"
+)
+
+// AllowOverwrite returns true if new state from a particular source is allowed
+// to overwrite existing state from another source
+func AllowOverwrite(existing, new Source) bool {
+	switch existing {
+
+	// Kubernetes state can be overwritten by everything except generated state
```
Comment (ianvernon) [inline pkg/source/source.go]: Nit: this should say `and unspec state`.

File: pkg/node/manager/manager.go [hunk 4]
```diff
@@ -355,6 +377,10 @@ func (m *Manager) NodeDeleted(n node.Node) {
 		return
 	}
 
+	for _, address := range entry.node.IPAddresses {
+		ipcache.IPIdentityCache.Delete(address.IP.String(), n.Source)
```
Comment (ianvernon) [inline pkg/node/manager/manager.go:381]: Is there a reason this cannot be placed in the `NodeDelete` call below on line 390? Similarly for the add case above, we could factor the ipcache update into the `NodeUpdate` part - this makes the `node` package unaware of the implementation of the datapath (e.g., the ipcache).

File: pkg/nodediscovery/nodediscovery.go [hunk 5]
```diff
@@ -210,10 +210,15 @@ func (n *NodeDiscovery) createCiliumNodeResource(conf Configuration) {
 
 	ciliumClient := k8s.CiliumClient()
 
-	nodeResource := &ciliumv2.CiliumNode{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: node.GetName(),
-		},
+	performUpdate := true
+	nodeResource, err := ciliumClient.CiliumV2().CiliumNodes().Get(node.GetName(), metav1.GetOptions{})
+	if err != nil {
```
Comment (ianvernon) [inline pkg/nodediscovery/nodediscovery.go:217]: Will the error here only result if the resource did not exist before? That is now I am interpreting the code - would be helpful to have a comment confirming this here.

File: daemon/k8s_watcher.go [hunk 6]
```diff
@@ -741,11 +745,115 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
...
+				&cilium_v2.CiliumNode{},
+				0,
+				cache.ResourceEventHandlerFuncs{
+					AddFunc: func(obj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumNode, metricCreate, valid, equal) }()
+						if ciliumNode, ok := obj.(*cilium_v2.CiliumNode); ok {
+							valid = true
+							n := node.ParseCiliumNode(ciliumNode)
+							if n.IsLocal() {
```
Comment (ianvernon) [inline daemon/k8s_watcher.go:773]: Why does locality of the node determine if it is equal?

File: daemon/k8s_watcher.go [hunk 7]
```diff
@@ -741,11 +745,115 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
...
+								d.nodeDiscovery.Manager.NodeUpdated(n)
+								d.K8sEventProcessed(metricCiliumNode, metricUpdate, true)
+								return nil
+							}, serializer.NoRetry)
+						}
+					},
+					DeleteFunc: func(obj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumNode, metricDelete, valid, equal) }()
+						deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
```
Comment (ianvernon) [inline daemon/k8s_watcher.go]: Would it be better to do the cast to `*cilium_v2.CiliumNode`  / check of this type first instead of `cache.DeletedFinalStateUnknown`? Depending on which case is more common `*cilium_v2.CiliumNode` vs. `cache.DeletedFinalStateUnknown`, we could be doing type checks more often than we have to. You could also switch on the type of the object itself as an alternative.

File: pkg/source/source.go [hunk 8]
```diff
@@ -48,6 +51,10 @@ func AllowOverwrite(existing, new Source) bool {
 	case Kubernetes:
 		return new != Generated && new != Unspec
 
+	// Custom-resource state can be overwritten by generated state
```
Comment (ianvernon) [inline pkg/source/source.go]: This comment appears to be wrong - it should be kvstore / local state / custom resource state.

File: daemon/k8s_watcher.go [hunk 9]
```diff
@@ -949,113 +947,6 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
 		}
 	}()
 
-	asyncControllers.Add(1)
```
Comment (ianvernon) [inline daemon/k8s_watcher.go:843]: I don't think we should remove this code because we are using `CiliumNode` for node discovery by default. it should be enabled / disabled based off of a configuration option that can be overridden by the user in case there are issues with using `CiliumNode` for node discovery by default.

File: operator/main.go [hunk 10]
```diff
@@ -223,11 +227,17 @@ func runOperator(cmd *cobra.Command) {
...
 
-	if synchronizeServices {
-		startSynchronizingServices()
-	}
+	if kvstoreEnabled() {
+		if synchronizeServices {
+			startSynchronizingServices()
+		}
+
+		if synchronizeNodes {
```
Comment (ianvernon) [inline operator/main.go]: Is there any concern behind moving starting of the the `NodeWatcher` before we have initialized the kvstore below? Changes to bootstrap ordering always catch my eye.

File: pkg/k8s/client_test.go (hunk 1 above)
Comment (tgraf) [reply pkg/k8s/client_test.go:151]: It was confusing me quite a bit because the local node was owned by LocalNode and AgentLocal cannot overwrite it. I then realized that the source is actually not used for this node instance which is why I've changed it to Unspec.

File: pkg/k8s/init.go (hunk 2 above)
Comment (tgraf) [reply pkg/k8s/init.go:88]: Correct

File: pkg/node/manager/manager.go (hunk 4 above)
Comment (tgraf) [reply pkg/node/manager/manager.go:381]: I agree. It's a bigger change though. It was unclear to me whether ipcache should be considered Linux specific or not. Assuming we would want to run tests with a mocked datapath, the ipcache updates would probably still be wanted but then mocked with a dummy BPF backend. Right now, the next level of abstraction beyond the node manager is an OS-specific datapath implementation.

I've opted for the simplest change into the latest common abstraction layer that is not OS specific yet.

File: daemon/k8s_watcher.go (hunk 6 above)
Comment (tgraf) [reply daemon/k8s_watcher.go:773]: So it will be counted as an unnecessary received update. We can also remove it, it's purely a metrics accounting thing.

File: daemon/k8s_watcher.go (hunk 7 above)
Comment (tgraf) [reply daemon/k8s_watcher.go]: I'm not sure what is more common, @aanm. This is only hit on node delete which is *not* (edit) a common operation anyway.

EDIT: added not

File: daemon/k8s_watcher.go (hunk 9 above)
Comment (tgraf) [reply daemon/k8s_watcher.go:843]: The cost is very high. What use case would justify using regular nodes instead of CiliumNodes?

File: operator/main.go (hunk 10 above)
Comment (tgraf) [reply operator/main.go]: I did not find any, otherwise, I wouldn't have done this change.

Comment (tgraf) [reply operator/main.go]: Noticed that unlike the service code, the node init code does not run async

Comment (tgraf) [reply operator/main.go]: Moved the code to after kvstore.Setup()

File: daemon/k8s_watcher.go (hunk 6 above)
Comment (tgraf) [reply daemon/k8s_watcher.go:773]: Removed it

File: pkg/nodediscovery/nodediscovery.go (hunk 5 above)
Comment (tgraf) [reply pkg/nodediscovery/nodediscovery.go:217]: There can be other errors but I'm not sure what would be a better error resolution. A blind `Delete()`? Maybe we should make it fail hard so the pod itself restarts.

Comment (tgraf) [reply pkg/nodediscovery/nodediscovery.go:217]: Changed to `Fatal()` logs to force retrying via pod restart

File: pkg/k8s/client_test.go (hunk 1 above)
Comment (ianvernon) [reply pkg/k8s/client_test.go:151]: Got it, thanks!

File: pkg/node/manager/manager.go (hunk 4 above)
Comment (ianvernon) [reply pkg/node/manager/manager.go:381]: OK, makes sense. We can punt this as a refactor - cc: @joestringer what do you think?

File: pkg/nodediscovery/nodediscovery.go (hunk 5 above)
Comment (ianvernon) [reply pkg/nodediscovery/nodediscovery.go:217]: I'm ambivalent. I just wasn't sure about the semantics behind the returned error.

File: daemon/k8s_watcher.go (hunk 9 above)
Comment (ianvernon) [reply daemon/k8s_watcher.go:843]: My thought is that we are moving to a mode which is only truly able to be tested via deployment in a cluster / integration testing, so having an easy way to fall back on existing behavior which we have tested before in case there is a bug by using `CiliumNode` is preferable than just deleting the old code outright. Once we have confidence in the new `CiliumNodes` mode we could delete the old K8s `Node` resource handling.

Comment (tgraf) [reply daemon/k8s_watcher.go:843]: I don't think we have a safe fall back other than kvstore and that still exists. kvstore will still take precedence as it did before. Node discovery on standard k8s resources is not really an alternative as it does not provide encryption information.

**Comment by tgraf** [comment]
test-me-please

File: daemon/k8s_watcher.go (hunk 9 above)
Comment (ianvernon) [reply daemon/k8s_watcher.go:843]: OK, makes sense - it's an all-or-nothing thing since encryption is involved.

**Review by ianvernon — APPROVED** [review]

File: daemon/k8s_watcher.go (hunk 9 above)
Comment (tgraf) [reply daemon/k8s_watcher.go:843]: Given the struggle we had with etcd-operator. My proposal is to focus all testing this week on gaining as much confidence as possible with this as we will rely on it replacing etcd-operator for simplicity while we refer to external etcd as the go-to scale option as we did before.

**Comment by tgraf** [comment]
CI: green build.

One local change not pushed yet:
 - Missing encryption key derival based on CiliumEndpoint CRD as pointed out by @jrfastab via Slack. The current code always uses the local node key.

**Comment by tgraf** [comment]
Rebased to include #8573 and to push change referred to by comment https://github.com/cilium/cilium/pull/8627#issuecomment-513972844

**Comment by tgraf** [comment]
test-me-please

File: daemon/k8s_watcher.go [hunk 11]
```diff
@@ -850,6 +853,102 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
...
+			close(isConnected)
+
+			log.Info("Connected to key-value store, stopping CiliumEndpoint watcher")
+
+			d.k8sAPIGroups.removeAPI(k8sAPIGroupCiliumEndpointV2)
+			// Create a new node controller when we are disconnected with the
+			// kvstore
+			<-kvstore.Client().Disconnected()
+
+			log.Info("Disconnected from key-value store, restarting CiliumEndpoint watcher")
```
Comment (joestringer) [inline daemon/k8s_watcher.go:941]: Should this code be protected by the `disable-endpoint-crd` option, to ensure that we don't cause trouble in a high-scale cluster whenever kvstore connectivity hits a connectivity blip?

File: pkg/nodediscovery/nodediscovery.go [hunk 12]
```diff
@@ -183,19 +183,21 @@ func (n *NodeDiscovery) StartDiscovery(nodeName string, conf Configuration) {
...
-				DoFunc: func(ctx context.Context) error {
-					err := n.Registrar.UpdateLocalKeySync(&n.LocalNode)
-					if err != nil {
-						log.WithError(err).Error("Unable to propagate local node change to kvstore")
-					}
-					return err
-				},
-			})
-	}()
+	if option.Config.KVStore != "" {
```
Comment (joestringer) [inline pkg/nodediscovery/nodediscovery.go:186]: This uses different logic from the daemon to determine if the kvstore is enabled, is that OK?

**Review by joestringer — COMMENTED** [review]
A few minor comments below.

File: daemon/k8s_watcher.go (hunk 7 above)
Comment (raybejjani) [reply daemon/k8s_watcher.go]: The convention in this file is to add and use `k8s.CopyObjTo*` functions, and then try twice: https://github.com/cilium/cilium/blob/2e9bb515eb87439df4dd96936e7d3d92d07e0299/daemon/k8s_watcher.go#L1000-L1013

I did violate this recently, however, because it would create an import loop.

**Review by raybejjani — APPROVED** [review]

File: pkg/nodediscovery/nodediscovery.go (hunk 12 above)
Comment (raybejjani) [reply pkg/nodediscovery/nodediscovery.go:186]: oh, I thought the daemon code was also switched out for this

File: daemon/k8s_watcher.go (hunk 11 above)
Comment (tgraf) [reply daemon/k8s_watcher.go:941]: Do you mean to not even attempt to install the watcher in the first place if `disable-endpoint-crd` is set?

File: pkg/nodediscovery/nodediscovery.go (hunk 12 above)
Comment (tgraf) [reply pkg/nodediscovery/nodediscovery.go:186]: I've based on this:
```
        if option.Config.KVStore == "" {
                log.Info("Skipping kvstore configuration")
        } else {
                d.initKVStore()
        }
```

Do you have something specific in mind?

File: daemon/k8s_watcher.go (hunk 7 above)
Comment (tgraf) [reply daemon/k8s_watcher.go]: I must have copied it from that bad example ;-)

Fixed.

File: daemon/k8s_watcher.go (hunk 11 above)
Comment (tgraf) [reply daemon/k8s_watcher.go:941]: I see that we are already automatically enabling the endpoint CRD when the kvstore is disabled so we should be all good?

**Comment by tgraf** [comment]
CI: green build

Pushing to address Ray's comment

**Comment by tgraf** [comment]
test-me-please

File: daemon/k8s_watcher.go (hunk 11 above)
Comment (joestringer) [reply daemon/k8s_watcher.go:941]: Right, I was suggesting to avoid this watcher codepath entirely if users are running with endpoint CRDs disabled. If that's already the case then feel free to ignore the suggestion.
